File: ising-sim/src/sim/monteCarlo.ts
import { beta, type SimConfig } from "./config";
import type { Geometry } from "./geometry";
import type { Interaction } from "./interaction";
import type { Lattice } from "./lattice";
import { SeededRng } from "./rng";

export interface SweepStats {
  attempts: number;
  acceptances: number;
}

export function acceptanceRate(stats: SweepStats): number {
  return stats.attempts === 0 ? 0 : stats.acceptances / stats.attempts;
}

export function sweep(
  lattice: Lattice,
  geometry: Geometry,
  interaction: Interaction,
  config: SimConfig,
  rng: SeededRng,
): SweepStats {
  const numSites = geometry.numSites();
  const inverseTemperature = beta(config);
  const stats: SweepStats = { attempts: 0, acceptances: 0 };

  for (let attempt = 0; attempt < numSites; attempt += 1) {
    const site = rng.nextInt(numSites);
    const delta = interaction.deltaEnergy(lattice, site, geometry, config.field);
    stats.attempts += 1;

    if (delta <= 0 || rng.nextFloat() < Math.exp(-inverseTemperature * delta)) {
      lattice.flip(site);
      stats.acceptances += 1;
    }
  }

  return stats;
}

export function runSweeps(
  lattice: Lattice,
  geometry: Geometry,
  interaction: Interaction,
  config: SimConfig,
  sweeps: number,
  seed?: number,
): SweepStats {
  const rng =
    seed === undefined
      ? new SeededRng(Date.now() >>> 0)
      : SeededRng.fromSeed(seed);
  const total: SweepStats = { attempts: 0, acceptances: 0 };

  for (let sweepIndex = 0; sweepIndex < sweeps; sweepIndex += 1) {
    const stats = sweep(lattice, geometry, interaction, config, rng);
    total.attempts += stats.attempts;
    total.acceptances += stats.acceptances;
  }

  return total;
}

File: ising-sim/backend/src/monte_carlo.rs
use crate::config::SimConfig;
use crate::geometry::Geometry;
use crate::interaction::Interaction;
use crate::lattice::Lattice;
use rand::Rng;
use rand::SeedableRng;
use rand::rngs::StdRng;

#[derive(Debug, Default, Clone, Copy)]
pub struct SweepStats {
    pub attempts: u64,
    pub acceptances: u64,
}

impl SweepStats {
    pub fn acceptance_rate(&self) -> f64 {
        if self.attempts == 0 {
            0.0
        } else {
            self.acceptances as f64 / self.attempts as f64
        }
    }
}

pub fn sweep(
    lattice: &mut Lattice,
    geometry: &dyn Geometry,
    interaction: &dyn Interaction,
    config: &SimConfig,
    rng: &mut StdRng,
) -> SweepStats {
    let num_sites = geometry.num_sites();
    let beta = config.beta();
    let mut stats = SweepStats::default();

    for _ in 0..num_sites {
        let site = rng.random_range(0..num_sites);
        let delta = interaction.delta_energy(lattice, site, geometry, config.field);
        stats.attempts += 1;

        if delta <= 0.0 || rng.random::<f64>() < (-beta * delta).exp() {
            lattice.flip(site);
            stats.acceptances += 1;
        }
    }

    stats
}

pub fn run_sweeps(
    lattice: &mut Lattice,
    geometry: &dyn Geometry,
    interaction: &dyn Interaction,
    config: &SimConfig,
    sweeps: u64,
    seed: Option<u64>,
) -> SweepStats {
    let mut rng = match seed {
        Some(seed) => StdRng::seed_from_u64(seed),
        None => StdRng::from_os_rng(),
    };

    let mut total = SweepStats::default();
    for _ in 0..sweeps {
        let stats = sweep(lattice, geometry, interaction, config, &mut rng);
        total.attempts += stats.attempts;
        total.acceptances += stats.acceptances;
    }
    total
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geometry::Square2DOpen;
    use crate::interaction::NearestNeighbor;

    #[test]
    fn negative_delta_energy_always_accepts() {
        let geometry = Square2DOpen::new(2, 2);
        let mut lattice = Lattice::try_from_spins(vec![1, -1, -1, 1]).unwrap();
        let interaction = NearestNeighbor::new(1.0);
        let config = SimConfig {
            temperature: 0.1,
            ..Default::default()
        };
        let mut rng = StdRng::seed_from_u64(42);

        let before = lattice.spins().to_vec();
        let stats = sweep(
            &mut lattice,
            &geometry,
            &interaction,
            &config,
            &mut rng,
        );

        assert!(stats.acceptances > 0);
        assert_ne!(before, lattice.spins());
    }
}
